posthog/user_permissions.py

```diff
@@ -0,0 +1,266 @@
+from functools import cached_property
+from typing import Any, Dict, List, Optional, cast
+from uuid import UUID
+
+from posthog.constants import AvailableFeature
+from posthog.models import Dashboard, DashboardTile, Insight, Organization, OrganizationMembership, Team, User
+
+
+class UserPermissions:
+    """
+    Class responsible for figuring out user permissions in an efficient manner.
+
+    Generally responsible for the following tasks:
+    1. Calculating whether a user has access to the current team
+    2. Calculating whether a user has access to other team(s)
+    3. Calculating permissioning of a certain object (dashboard, insight) in the team
+
+    Note that task 3 depends on task 1, so for efficiency sake the class _generally_
+    expects the current team/organization to be passed to it and will use it to skip certain
+    lookups.
+    """
+
+    def __init__(self, user: User, team: Optional[Team] = None):
+        self.user = user
+        self._current_team = team
+
+        self._tiles: Optional[List[DashboardTile]] = None
+        self._team_permissions: Dict[int, UserTeamPermissions] = {}
+        self._dashboard_permissions: Dict[int, UserDashboardPermissions] = {}
+        self._insight_permissions: Dict[int, UserInsightPermissions] = {}
+
+    @cached_property
+    def current_team(self) -> "UserTeamPermissions":
+        if self._current_team is None:
+            raise ValueError("Cannot call .current_team without passing it to UserPermissions")
+
+        return UserTeamPermissions(self, self._current_team)
+
+    def team(self, team: Team) -> "UserTeamPermissions":
```

Comment:
nit: `team` and `dashboard` don't return a `team` or `dashboard`

Should they be `team_permissions` and `dashboard_permissions`?